starport/services/network/join_test.go

```diff
@@ -0,0 +1,370 @@
+		mock.AnythingOfType("*types.MsgRequestAddValidator"),
+	).Return(testutil.NewResponse(&launchtypes.MsgRequestAddValidatorResponse{
+		RequestID:    TestGenesisValidatorRequestID,
+		AutoApproved: false,
+	}), nil)
+	networkClientMock.On(
+		"BroadcastTx",
+		testutil.AccountName,
+		mock.AnythingOfType("*types.MsgRequestAddAccount"),
+	).Return(testutil.NewResponse(&launchtypes.MsgRequestAddAccountResponse{
+		RequestID:    TestAccountRequestID,
+		AutoApproved: false,
+	}), nil)
+	return Network{
+		cosmos:      networkClientMock,
+		account:     testutil.GetTestAccount(),
+		launchQuery: launchQueryMock,
+	}
+}
+
+func TestJoin(t *testing.T) {
+	tmp := t.TempDir()
+	gentx := testutil.NewGentx(testutil.Address, TestDenom, TestAmountString, "")
+	gentxPath, err := gentx.SaveTo(tmp)
+	assert.Nil(t, err)
+	genesis := testutil.NewGenesis(testutil.TestChainChainID)
+	genesisPath, err := genesis.SaveTo(tmp)
+	assert.Nil(t, err)
+
+	network := stubNetworkForJoin()
+	chain := testutil.NewChainMock(testutil.WithGenesisPath(genesisPath), testutil.WithDefaultGentxPath(gentxPath))
+
+	joinErr := network.Join(context.Background(), chain, testutil.TestLaunchID, testutil.TestPublicAddress, "")
+	require.Nil(t, joinErr)
+	chain.AssertNumberOfCalls(t, "NodeID", 1)
+	chain.AssertNumberOfCalls(t, "GenesisPath", 1)
+	chain.AssertNumberOfCalls(t, "DefaultGentxPath", 1)
+	network.launchQuery.(*mocks.LaunchClient).AssertNumberOfCalls(t, "GenesisValidator", 1)
+	network.cosmos.(*mocks.CosmosClient).AssertNumberOfCalls(t, "BroadcastTx", 1)
+}
+
+func TestJoinWithCustomGentx(t *testing.T) {
+	tmp := t.TempDir()
+	gentx := testutil.NewGentx(testutil.Address, TestDenom, TestAmountString, "")
+	gentxJson, err := gentx.JSON()
+	assert.Nil(t, err)
+	gentxPath, err := gentx.SaveTo(tmp)
+	assert.Nil(t, err)
+	genesis := testutil.NewGenesis(testutil.TestChainChainID)
+	genesisPath, err := genesis.SaveTo(tmp)
+	assert.Nil(t, err)
+
+	network := stubNetworkForJoin()
+	chain := testutil.NewChainMock(testutil.WithGenesisPath(genesisPath))
+
+	joinErr := network.Join(context.Background(), chain, testutil.TestLaunchID, testutil.TestPublicAddress, gentxPath)
+	require.Nil(t, joinErr)
+	chain.AssertNumberOfCalls(t, "NodeID", 1)
+	chain.AssertNumberOfCalls(t, "GenesisPath", 1)
+	chain.AssertNumberOfCalls(t, "DefaultGentxPath", 0)
+	network.launchQuery.(*mocks.LaunchClient).AssertNumberOfCalls(t, "GenesisValidator", 1)
+	network.cosmos.(*mocks.CosmosClient).AssertCalled(t,
+		"BroadcastTx",
+		mock.Anything,
+		&launchtypes.MsgRequestAddValidator{
+			Creator:        testutil.Address,
+			LaunchID:       testutil.TestLaunchID,
+			ValAddress:     testutil.Address,
+			GenTx:          gentxJson,
+			ConsPubKey:     []byte{},
+			SelfDelegation: sdk.NewCoin(TestDenom, sdk.NewInt(TestAmountInt)),
+			Peer: launchtypes.Peer{
+				Id: testutil.TestNodeID,
+				Connection: &launchtypes.Peer_TcpAddress{
+					TcpAddress: testutil.TestPublicAddress,
+				},
+			},
+		},
+	)
+}
```

Comment:
Shouldn't we run `Mock.AssertExpectations` in the end of all tests?

Reply:
AsserrtExpectations asserts that everything specified with On() was called exact number of times. There are definitely  cases where this method could be used.